phare/pp/diagnostics.py

```diff
@@ -0,0 +1,234 @@
+# formatted with black code formatter
+
+import os, h5py, phare.pharein as ph, itertools
+from .diagnostic.patch_data import _EM, _Fluid, _Particle, Particles
+from .diagnostic.patch import Patch, PatchLevel
+
+
+class H5FileRegexer:
+    """Used to Validates h5 input file to ignore files which are not to be read/extracted"""
+
+    @staticmethod
+    def ignore_population_name_in(h5FilesPerType):
+        import re
+
+        pop = "${POP}"
+        regexing = {Type: [] for Type in h5FilesPerType.keys()}
+        for Type, files in h5FilesPerType.items():
+            for file in files:
+                if pop not in file:
+                    continue
+                bits = file.split(pop)
+                left, right = bits[0], bits[1]
+                regexing[Type].append(
+                    (file, re.compile("^" + left + "(.*)" + right + "$"))
+                )
+        return regexing
+
+
+class Diagnostic:
+    REFINEMENT_RATIO = 2
+    """12 decimal places is arbitrary, but hopefully sufficient"""
+    DL_PRECISION = 12
+
+    """List of all possible input HDF5 Files"""
+    h5FilesPerType = {
+        _EM: ["EM_B.h5", "EM_E.h5"],
+        _Fluid: [
+            "ions_bulkVelocity.h5",
+            "ions_density.h5",
+            "ions_pop_ions_${POP}_density.h5",
+            "ions_pop_ions_${POP}_flux.h5",
+        ],
+        _Particle: [
+            "ions_pop_ions_${POP}_domain.h5",
+            "ions_pop_ions_${POP}_patchGhost.h5",
+            "ions_pop_ions_${POP}_levelGhost.h5",
+        ],
+    }
+    h5Files = [f for files in itertools.chain(h5FilesPerType.values()) for f in files]
+
+    """Compiled Regexs to ignore population name in HDF5 input files"""
+    h5FilesRegexesPerType = H5FileRegexer.ignore_population_name_in(h5FilesPerType)
+    h5FilesRegexes = {
+        r[0]: r[1]
+        for rgxs in itertools.chain(h5FilesRegexesPerType.values())
+        for r in rgxs
+    }
+
+    def __init__(self, sim: ph.Simulation, file):
+        def truncate(number, digits) -> float:
+            import math
+
+            stepper = 10.0 ** digits
+            return math.trunc(stepper * number) / stepper
+
+        self.sim = sim
+        self.file = file
+        self.patch_data_type = Diagnostics.get_type_from_file(file)
+        self.hifile = h5py.File(file, "r")
+        self.levels = {}
+        self.dl = [truncate(x, Diagnostic.DL_PRECISION) for x in sim.dl]
```

Comment:
I don't understand why the diagnostic should have the `dl`, that's the coarsest mesh size and not a property of the diag. I would think the diag should not have that.

Reply:
it's a file attribute of the diagnostic h5 file. no associated with any patch 

will be here: https://github.com/PHAREHUB/PHARE/blob/master/src/diagnostic/detail/h5writer.h#L195

I would think these are "diagnostic" attributes.

Reply:
does https://github.com/PHAREHUB/PHARE/pull/106/files#r402554424 resolve this?